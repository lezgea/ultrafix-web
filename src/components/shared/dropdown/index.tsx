"use client";

import React, { useState, ReactNode, useRef, useEffect } from 'react';

interface IDropdownProps {
    content?: ReactNode;
    children: ReactNode;
}

export const Dropdown: React.FC<IDropdownProps> = ({ content, children }) => {
    const [isOpen, setIsOpen] = useState<boolean>(false);
    const dropdownRef = useRef<HTMLDivElement>(null);

    const toggleDropdown = () => {
        setIsOpen((prevState) => !prevState);
    };

    const handleClickOutside = (event: MouseEvent) => {
        if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
            setIsOpen(false);
        }
    };

    useEffect(() => {
        document.addEventListener('mousedown', handleClickOutside);
        return () => {
            document.removeEventListener('mousedown', handleClickOutside);
        };
    }, []);

    return (
        <div className="relative inline-block text-left" ref={dropdownRef}>
            <div onClick={toggleDropdown} aria-expanded={isOpen} role="button" tabIndex={0} onKeyDown={(e) => {
                if (e.key === 'Enter' || e.key === ' ') toggleDropdown();
            }}>
                {children} {/* Renders children element */}
            </div>
            {isOpen && (
                <div onClick={toggleDropdown} className={`origin-top-right absolute right-0 mt-2 rounded-xl shadow-lg bg-white ring-1 ring-black ring-opacity-5 p-2 transition-all duration-500 transform ${isOpen ? 'scale-100 opacity-100' : 'scale-10 opacity-0'}`}>
                    {content}
                </div>
            )}
        </div>
    );
};
