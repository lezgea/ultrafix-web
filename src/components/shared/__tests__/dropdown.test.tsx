import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Dropdown } from '../dropdown';


describe('Dropdown Component', () => {
    const content = <div>Dropdown Content</div>;
    const trigger = <button>Open Dropdown</button>;

    it('should render the children and content correctly', () => {
        const { container, asFragment } = render(
            <Dropdown content={content}>
                {trigger}
            </Dropdown>
        );

        expect(screen.getByText('Open Dropdown')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should toggle the dropdown content on click', () => {
        render(
            <Dropdown content={content}>
                {trigger}
            </Dropdown>
        );

        const triggerElement = screen.getByText('Open Dropdown');

        expect(screen.queryByText('Dropdown Content')).not.toBeInTheDocument();
        fireEvent.click(triggerElement);
        expect(screen.getByText('Dropdown Content')).toBeInTheDocument();
        fireEvent.click(triggerElement);
        expect(screen.queryByText('Dropdown Content')).not.toBeInTheDocument();
    });

    it('should close the dropdown when clicking outside', () => {
        render(
            <Dropdown content={content}>
                {trigger}
            </Dropdown>
        );

        const triggerElement = screen.getByText('Open Dropdown');

        fireEvent.click(triggerElement);
        expect(screen.getByText('Dropdown Content')).toBeInTheDocument();
        fireEvent.mouseDown(document);
        expect(screen.queryByText('Dropdown Content')).not.toBeInTheDocument();
    });

    it('should match snapshot when dropdown is open', () => {
        const { asFragment } = render(
            <Dropdown content={content}>
                {trigger}
            </Dropdown>
        );

        fireEvent.click(screen.getByText('Open Dropdown'));
        expect(asFragment()).toMatchSnapshot();
    });
});
