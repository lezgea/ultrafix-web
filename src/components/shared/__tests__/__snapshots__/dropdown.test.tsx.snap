// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component should match snapshot when dropdown is open 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block text-left"
  >
    <div
      aria-expanded="true"
      role="button"
      tabindex="0"
    >
      <button>
        Open Dropdown
      </button>
       
    </div>
    <div
      class="origin-top-right absolute right-0 mt-2 rounded-xl shadow-lg bg-white ring-1 ring-black ring-opacity-5 p-2 transition-all duration-500 transform scale-100 opacity-100"
    >
      <div>
        Dropdown Content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown Component should render the children and content correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block text-left"
  >
    <div
      aria-expanded="false"
      role="button"
      tabindex="0"
    >
      <button>
        Open Dropdown
      </button>
       
    </div>
  </div>
</DocumentFragment>
`;
